Collect node custom attributes from hawkular during refresh

zgalor opened the pull request:
Reading metrics from hawkular tagged with 'miq_metric=true' and saving them as custom attributes in the node they were collected from. This is done in ems refresh process for kubernetes and openshift.

![additional_attributes](https://cloud.githubusercontent.com/assets/20948594/22077155/a95943c4-ddbb-11e6-97fb-6d712530952e.png)

cc @simon3z 
@miq-bot add_label providers/containers

zgalor commented:
@yaacov PTAL

simon3z commented:
@moolitayer @zeari @cben can you review?

simon3z commented:
@zgalor I would expected this to be part of `providers/kubernetes/container_manager/refresh_parser.rb` and not metrics collection.

What client you use to gather information shouldn't affect *where* you do the collection.

moolitayer commented:
@simon3z There is some sense in doing this collection under the new provider. That way we would not interfere with upstream customers out of the box.

simon3z commented:
@moolitayer there are few things that we want to do in the hawkular provider but this one seems to really belong to inventory refresh on kubernetes/openshift.

@miq-bot assign zgalor

zgalor commented:
@miq-bot add_label wip

zgalor commented on app/models/manageiq/providers/kubernetes/container_manager/refresher_mixin.rb:
```diff
@@ -21,6 +21,18 @@ def fetch_entities(client, entities)
         end
+
+        def fetch_ems_metrics(ems)
+          hawk = ManageIQ::Providers::Kubernetes::ContainerManager::MetricsCapture::HawkularClient.new(ems)
```
@yaacov here I am creating a new HawkularClient for the ems, even though the ems already has a hawkular client for metrics. This is done once in every refresh for each provider. 
Is that a good practice? Do you see a better way to do it?

zgalor commented:
I removed the previous commit - collecting string metrics as part of metric collection and re-implemented as part of refresh.
I also renamed the PR and updated the comment.
It's still a wip because I need to add specs, but I'd like to get some reviews.

@moolitayer @enoodle @cben @yaacov

enoodle commented on app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -395,6 +395,19 @@ def save_container_volumes_inventory(container_group, hashes, target = nil)
     store_ids_for_new_records(container_group.container_volumes, hashes, :name)
   end
 
+  def save_custom_attributes_inventory(entity, hashes, target = nil)
```
maybe we can change `save_custom_attribute_attribute_inventory` to use this function

zgalor replied:
I suggest you refactor this in a different PR

enoodle replied:
Yes. I will rephrase: What a cool function! Lets rewrite `save_custom_attribute_attribute_inventory` with it in another PR!

enoodle commented on app/models/manageiq/providers/kubernetes/container_manager/refresher_mixin.rb:
```diff
@@ -21,6 +21,18 @@ def fetch_entities(client, entities)
+          keys = hawk.strings.query(:miq_metric => true)
+
+          metrics = {}
+          keys.each do |k|
```
maybe you can use `each_with_objecct` here? instead of defining the metrcis hash and then returning it:
```
keys.each_with_object({|) do |k, metrics|
  .....
end
```

zgalor replied:
:+1:

enoodle commented on app/models/manageiq/providers/openshift/container_manager/refresher.rb:
```diff
@@ -20,6 +20,7 @@ def parse_legacy_inventory(ems)
           fetch_entities(openshift_client, OPENSHIFT_ENTITIES)
         end
         entities = openshift_entities.merge(kube_entities)
+        entities[:metrics] = fetch_ems_metrics(ems)
```
Is this only collecting metrics for nodes or also for other entities?

zgalor commented on app/models/manageiq/providers/openshift/container_manager/refresher.rb:
```diff
@@ -12,6 +12,16 @@ class ContainerManager::Refresher < ManageIQ::Providers::BaseManager::Refresher
       ]
 
+      def fetch_ems_metrics(ems)
+        hawk = ManageIQ::Providers::Kubernetes::ContainerManager::MetricsCapture::HawkularClient.new(ems)
```
@yaacov here I am creating a new HawkularClient for the ems, even though the ems already has a hawkular client for metrics. This is done once in every refresh for each provider.
Is that a good practice? Do you see a better way to do it?

yaacov commented on app/models/manageiq/providers/openshift/container_manager/refresher.rb:
```diff
@@ -20,6 +32,7 @@ def parse_legacy_inventory(ems)
           fetch_entities(openshift_client, OPENSHIFT_ENTITIES)
         end
         entities = openshift_entities.merge(kube_entities)
+        entities[:metrics] = fetch_ems_metrics(ems) || {}
```
see above comment about naming the functions `fetch_ems_metrics`

you do not collect metrics [ e.g. cpu_usage, filesystem_usage, network_rate .... ] you collect inventory [ e.g. server versions ] calling it metrics is confusing in this context.

moolitayer commented:
@zgalor documenting my suggestions per your request, let me know if I forgot anything
- entities[:additional_attributes] will contain metrics from hawkular[1] 
- in refresh_parser add a generic function that for now is called only from container nodes (I think we called it `parse_additional_attributes`?)
- use fetch_path in that function to locate the node by name and add the attributes

[1] we could already store it on the node at that point, but I prefer to keep the node name parsing out of refresher.rb and in refresh_parser.rb

zgalor commented:
As agreed with @moolitayer and @yaacov:

1. The method reading hawkular inventory was renamed to: fetch_hawkular_inv
2. The key in inventory where hawk data is stored was renamed to "additional_attributes"
3. get_additional_attributes was added to parse additional attributes into nodes using store_path and fetch_path, instead of changing parse_node API.

Temporarily tests fail, until I fix the refresher spec.
Currently the code is split to 2 commits, but I'll merge it once I'm done.

zeari commented:
Since `custom attributes` also describe labels\selectors\annotations i think that its better to name these something else and add a `:has_many` in the model similar to labels

zgalor commented:
@zeari got a name suggestion?

zgalor commented:
@miq-bot remove_label wip

zgalor commented:
I added the final spec and removed the wip label.
Can you guys give a final look and/or a :+1: ?

@yaacov @moolitayer @enoodle 

All apart from @zeari 's comment regarding the naming

zeari commented:
> @zeari got a name suggestion?

Not really, can these just be lumped under `labels` with the kubernetes labels?

zgalor commented:
> Not really, can these just be lumped under labels with the kubernetes labels?

Not sure. These aren't labels. labels might have a different purpose - like for tagging and grouping.
@simon3z ?

zgalor commented:
I added to the refresher spec a recording that fails connecting to hawkular and still records the inventory, without the hawk node custom attributes.

simon3z commented:
> Not really, can these just be lumped under labels with the kubernetes labels?
> Not sure. These aren't labels. labels might have a different purpose - like for tagging and grouping.
> @simon3z ?

@zgalor you need a naming suggestion for the section of the nodes custom attributes?
Maybe something like `nodes_attributes` ? or `node_extra_attributes`.

zgalor commented:
@zeari @simon3z I used the generic name additional_attributes so the same code can be later used for other entities (much like labels).

I also added a screenshot in the description.
Please make final reviews/approval so this can be merged.

simon3z commented:
> @zeari @simon3z I used the generic name additional_attributes so the same code can be later used for other entities (much like labels).

@zgalor I am not sure `additional` is descriptive enough. I have the feeling we can find something better... (we won't block on this but it would be great to improve the naming).

zgalor commented:
> @zgalor I am not sure additional is descriptive enough. I have the feeling we can find something better... (we won't block on this but it would be great to improve the naming).
@simon3z I wanted the inventory parsing to be transparent instead of specific to node so it can be used for other entities. Does "**dynamic_attributes**" sound better?

simon3z commented:
LGTM :+1:  cc @chessbyte 
Assigning to @blomquisg for review/merge.

@miq-bot assign blomquisg

blomquisg commented on app/models/ems_refresh/save_inventory_container.rb:
```diff
@@ -395,6 +396,19 @@ def save_container_volumes_inventory(container_group, hashes, target = nil)
     store_ids_for_new_records(container_group.container_volumes, hashes, :name)
   end
 
+  def save_additional_attributes_inventory(entity, hashes, target = nil)
```
I'd like to see this merge with the [save_custom_attributes_inventory](https://github.com/ManageIQ/manageiq/blob/9db73c860bf9a5cf6ad030db881177ed3a47035d/app/models/ems_refresh/save_inventory.rb#L285) method since `additional_attributes` is really just an [alias for `custom_attributes`](https://github.com/ManageIQ/manageiq/blob/9db73c860bf9a5cf6ad030db881177ed3a47035d/app/models/ems_refresh/save_inventory.rb#L285).  But, this has two major differences:

1.  `additional_attributes` vs. `custom_attributes` in the call to `save_inventory_multi`
2. this version calls `store_ids_for_new_records` whereas `save_custom_attributes_inventory` does not ... and I'm not sure why.

Fryguy replied:
I agree...the model should just be `has_many :custom_attributes` instead of trying to create some special naming.  If store_ids_for_new_records is needed, there's no reason it can't be added to the existing save_custom_attributes_inventory

Fryguy replied:
On that note though...is custom_attributes the right place for this or should this be in advanced_settings?  It depends on the what is going to be store here in the refresh.

The difference is that custom_attributes are things that people manipulate and change, particularly through manageiq.  advanced_settings are extended metadata that is mostly for informational purposes.

zgalor replied:
@Fryguy this is data that we collect daily from the openshift cluster such as rpm version of openshift in each node, and then create a policy that verifies the attributed against the expected version. Could also be certificate fingerprint. Not intended to be manipulated through manageIQ, but to be pushed and changed externally.

@blomquisg I initially created these as regular custom attributes, but as @zeari noted custom_attributes are very generic, and we might want to separate those specific attributes (like we do for labels\selectors\annotations)

zgalor replied:
@Fryguy will advanced_settings show on the Node/Provider page? Will I be able to use them in conditions?

Fryguy replied:
I believe both, yes, but @gtanzillo can you verify?

blomquisg replied:
Hrm, looking at the two tables, I think @Fryguy is backwards on this.  `advanced_settings` has `default_value`, `min`, `max`, and `read_only` attributes.  Which to me looks like something that might be used to build a UI for manipulating in ManageIQ.

Whereas `custom_attributes` has just `section`, `name`, `value`, and `description`.  Which, to me means it's data we simply collect and display.

So, it looks like `custom_attributes` is where this should stay.

However, I still think it would be better to reuse the `custom_attributes` directly (so you could reuse the existing `save_custom_attributes_inventory` method, maybe with a the small change of adding `store_ids_for_new_records` to the current implementation).  Then, you could always group these Hawkular based settings with the `section` attribute of `custom_attributes`.  And, if needed, you could have a special method called `additional_attributes` that grabs these Hawkular settings.

That seems better than trying to circumvent the current inventory just to have a collection attribute new name on `container_node`.

@zeari, @zgalor, @Fryguy thoughts?

zgalor replied:
I changed the code to reuse existing `save_custom_attribute_attribute_inventory`.
I kept the `additional_attributes` alias same way we do for `labels`, for easier access and management and to differentiate those custom_attributes from others.

@blomquisg PTAL

Fryguy replied:
> advanced_settings has default_value, min, max, and read_only attributes.

that's because it was built from vmware which gave us those values.  advanced_settings there were modifiable, but not common to modify and definitely not modified by us.  custom_attributes was closer to like amazon tags or openshift labels, where it was quite common to change/add/delete them.  So if this is what these hawkular settings are, then it belongs in custom_attributes.

Fryguy replied:
:+1: I like this new change.

gtanzillo replied:
@Fryguy @zgalor Custom attributes are available in reporting and expressions but, currently, advanced settings are not. Well, they are but not in the way custom attributes are where the CustomAttribute#name appears as a reportable column. It's on our todo list, though.

blomquisg commented:
This is good @zgalor ... thanks for working with us on it!
